src/sema/types.rs

```diff
@@ -1556,7 +1560,11 @@ impl Type {
     /// which is not accounted for.
     pub fn solana_storage_size(&self, ns: &Namespace) -> BigInt {
         match self {
-            Type::Array(_, dims) if dims.first() == Some(&ArrayLength::Dynamic) => 4.into(),
+            Type::Array(_, dims) if dims.contains(&ArrayLength::Dynamic) => {
+                let size = dynamic_array_size(dims);
+                // A pointer is four bytes on Solana
+                size * 4
```

Comment:
Wouldn't this be the size + 4 times the number of dynamic arrays?

Reply:
As I understand, an array `uint8 [2][][3]` means a fixed-length array of pointers that contains three elements. Each pointer points to `uint8[2]`. The size will only be `3*pointer_size`. Now if the type was `uint8 [2][][][3]`, we'll still have a fixed-array of pointers that contains three elements. Each pointer points to a `uint8[2][]`. The size will continue to be `3*pointer_size`.

The function calculates the minimum size we need for storage. As soon as we allocate elements, more space will be necessary.

Reply:
I see, thanks.

What you explained here would look good in a comment to the source code (or even in the doc comment of this function)

Reply:
This is explained right above `dynamic_array_size` between lines 2172 and 2176.